### PR TITLE
rendergraph: Port perl script to Python and add pre-commit hook

### PR DESCRIPTION
**Pull request by Holzhaus**

### CODE REVIEW

File: tools/rg_generate_shaders_gl.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+                        buffered_blank_line = False
+                    yield line
+                else:
+                    buffered_blank_line = True
+        elif in_shader_block == 1:
+            if line.rstrip() == "Contents:":
+                in_shader_block = 2
+        else:
+            if line.rstrip() == "Shader 1: GLSL 120 [Standard]":
+                in_shader_block = 1
```
Comment (acolombier) [inline tools/rg_generate_shaders_gl.py:91]: Very naive suggestion here, to help flattening the nested logic

```suggestion
        if in_shader_block == 2:
            if re.match(r"^\*\*", line):
                ok = True
            else:
                if not comment_added and not re.match(r"^#", line):
                    yield "//// GENERATED - EDITS WILL BE OVERWRITTEN"
                    comment_added = True
                if line:
                    if buffered_blank_line:
                        yield ""
                        buffered_blank_line = False
                    yield line
                else:
                    buffered_blank_line = True
        elif in_shader_block == 1 and line.rstrip() == "Contents:":
                in_shader_block = 2
        elif line.rstrip() == "Shader 1: GLSL 120 [Standard]":
                in_shader_block = 1
```

File: tools/rg_generate_shaders_gl.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+    raise OSError(f"Executable {executable_name!r} not found!")
+
+
+QSB_EXECUTABLE = find_executable(
+    "qsb",
+    additional_paths=[
+        "/usr/lib/qt6/bin",
+        "/lib/qt6/bin",
+        "/usr/local/lib/qt6/bin",
+    ],
```
Comment (acolombier) [inline tools/rg_generate_shaders_gl.py:44]: Should we use `$PATH` as well?
```suggestion
    additional_paths=[
        "/usr/lib/qt6/bin",
        "/lib/qt6/bin",
        "/usr/local/lib/qt6/bin",
    ] + [os.getenv("PATH", "").split(":")],
```

**Review by acolombier — COMMENTED** [review]
Thanks for looking into this! The python version is definitely more beefy but also should support wider system setup, as well as provide logging a command usage.

File: tools/rg_generate_shaders_gl.py (hunk 2 above)
Comment (Holzhaus) [reply tools/rg_generate_shaders_gl.py:44]: We call `shutil.which` up to two times in the funtion, the first time without setting the path argument which implicitly defaults to `$PATH`. So yeah, that is already supported.

Comment (acolombier) [reply tools/rg_generate_shaders_gl.py:44]: Ah good to know. Thanks for the lesson!

**Comment by acolombier** [comment]
I'm getting a segfault at startup

```
Thread 1 "mixxx" received signal SIGSEGV, Segmentation fault.
0x00005555563430eb in allshader::WaveformWidget::WaveformWidget (this=0x55558e6fcca0, parent=<optimised out>, type=WaveformWidgetType::RGB, group=..., options=...) at /home/antoine/dev/mixxx/src/waveform/widgets/allshader/waveformwidget.cpp:52
52	    if (m_pWaveformRendererSignal && m_pWaveformRendererSignal->supportsSlip()) {
(gdb) bt
#0  0x00005555563430eb in allshader::WaveformWidget::WaveformWidget(QWidget*, WaveformWidgetType::Type, QString const&, QFlags<allshader::WaveformRendererSignalBase::Option>)
    (this=0x55558e6fcca0, parent=<optimised out>, type=WaveformWidgetType::RGB, group=<optimised out>, options=...) at /home/antoine/dev/mixxx/src/waveform/widgets/allshader/waveformwidget.cpp:52
#1  0x0000555556268a56 in WaveformWidgetFactory::createAllshaderWaveformWidget(WaveformWidgetType::Type, WWaveformViewer*) (this=this@entry=0x55555c230cc0, type=type@entry=WaveformWidgetType::RGB, viewer=viewer@entry=0x55558abbef60)
    at /home/antoine/dev/mixxx/src/waveform/waveformwidgetfactory.cpp:1005
#2  0x0000555556268eb6 in WaveformWidgetFactory::createRGBWaveformWidget(WWaveformViewer*) (this=this@entry=0x55555c230cc0, viewer=viewer@entry=0x55558abbef60) at /home/antoine/dev/mixxx/src/waveform/waveformwidgetfactory.cpp:1063
#3  0x0000555556267305 in WaveformWidgetFactory::createWaveformWidget(WaveformWidgetType::Type, WWaveformViewer*) (this=this@entry=0x55555c230cc0, type=WaveformWidgetType::RGB, viewer=viewer@entry=0x55558abbef60)
    at /home/antoine/dev/mixxx/src/waveform/waveformwidgetfactory.cpp:1141
#4  0x0000555556267a90 in WaveformWidgetFactory::setWidgetTypeFromHandle(int, bool) (this=this@entry=0x55555c230cc0, handleIndex=handleIndex@entry=0, force=<optimised out>) at /home/antoine/dev/mixxx/src/waveform/waveformwidgetfactory.cpp:617
#5  0x0000555556266df4 in WaveformWidgetFactory::setWidgetTypeFromConfig() (this=0x55555c230cc0) at /home/antoine/dev/mixxx/src/waveform/waveformwidgetfactory.cpp:583
#6  0x00007ffff17ac023 in  () at /lib/x86_64-linux-gnu/libQt6Core.so.6
#7  0x000055555603e6f4 in MixxxMainWindow::skinLoaded() (this=0x7fffffffd6b0) at mixxx-lib_autogen/include/moc_mixxxmainwindow.cpp:272
#8  MixxxMainWindow::loadConfiguredSkin() (this=this@entry=0x7fffffffd6b0) at /home/antoine/dev/mixxx/src/mixxxmainwindow.cpp:1234
#9  0x000055555603b66a in MixxxMainWindow::initialize() (this=0x7fffffffd6b0) at /home/antoine/dev/mixxx/src/mixxxmainwindow.cpp:324
#10 0x00007ffff17ac023 in  () at /lib/x86_64-linux-gnu/libQt6Core.so.6
#11 0x00007ffff4677ada in QOpenGLWindow::resizeEvent(QResizeEvent*) () at /lib/x86_64-linux-gnu/libQt6OpenGL.so.6
#12 0x00007ffff1e10858 in QWindow::event(QEvent*) () at /lib/x86_64-linux-gnu/libQt6Gui.so.6
#13 0x0000555556344557 in OpenGLWindow::event(QEvent*) (this=0x353167223d646920, pEv=0x0) at /home/antoine/dev/mixxx/src/widget/openglwindow.cpp:70
#14 0x00007ffff4f7fd36 in QApplicationPrivate::notify_helper(QObject*, QEvent*) () at /lib/x86_64-linux-gnu/libQt6Widgets.so.6
#15 0x000055555602bfcf in MixxxApplication::notify(QObject*, QEvent*) (this=0x7fffffffd630, pTarget=0x55555baf1d40, pEvent=0x7fffffffd2f0) at /home/antoine/dev/mixxx/src/mixxxapplication.cpp:201
#16 0x00007ffff1753a58 in QCoreApplication::notifyInternal2(QObject*, QEvent*) () at /lib/x86_64-linux-gnu/libQt6Core.so.6
#17 0x00007ffff1dd0300 in QGuiApplicationPrivate::processExposeEvent(QWindowSystemInterfacePrivate::ExposeEvent*) () at /lib/x86_64-linux-gnu/libQt6Gui.so.6
#18 0x00007ffff1e13c8c in QWindowSystemInterface::sendWindowSystemEvents(QFlags<QEventLoop::ProcessEventsFlag>) () at /lib/x86_64-linux-gnu/libQt6Gui.so.6
#19 0x00007fffe925c86e in  () at /usr/lib/x86_64-linux-gnu/qt6/plugins/platforms/../../../libQt6XcbQpa.so.6
#20 0x00007ffff5d1bd3b in g_main_context_dispatch () at /lib/x86_64-linux-gnu/libglib-2.0.so.0
#21 0x00007ffff5d712b8 in  () at /lib/x86_64-linux-gnu/libglib-2.0.so.0
#22 0x00007ffff5d193e3 in g_main_context_iteration () at /lib/x86_64-linux-gnu/libglib-2.0.so.0
#23 0x00007ffff197beae in QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) () at /lib/x86_64-linux-gnu/libQt6Core.so.6
#24 0x00007ffff1760adb in QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) () at /lib/x86_64-linux-gnu/libQt6Core.so.6
#25 0x00007ffff175c0f3 in QCoreApplication::exec() () at /lib/x86_64-linux-gnu/libQt6Core.so.6
#26 0x0000555555b427e4 in (anonymous namespace)::runMixxx(MixxxApplication*, CmdlineArgs const&) (pApp=<optimised out>, args=<optimised out>) at /home/antoine/dev/mixxx/src/main.cpp:108
#27 main(int, char**) (argc=<optimised out>, argv=<optimised out>) at /home/antoine/dev/mixxx/src/main.cpp:243
```

**Comment by m0dB** [comment]
Thanks @Holzhaus !

**Pull request merged**